metrics/chronos_httpserver.nim

```diff
@@ -55,10 +76,367 @@ when defined(metrics):
+          true
+        else:
+          if data.http.state() == HttpServerState.ServerClosed:
+            await data.respond(ResponseError, "HTTP server is already closed")
+            true
+          else:
+            await data.http.closeWait()
+            await data.respond(ResponseOk, "")
+            false
+    except CatchableError as exc:
+      printError(exc.msg)
+      if not(isNil(data.http)):
+        await data.http.closeWait()
+      return false
+
+  proc asyncLoop(server: MetricsServerData) {.async.} =
+    var lastError = ""
+
+    proc cb(r: RequestFence): Future[HttpResponseRef] {.async.} =
+      if r.isOk():
+        let request = r.get()
+        try:
+          if request.uri.path == "/metrics":
+            {.gcsafe.}:
+              # Prometheus will drop our metrics in surprising ways if we give
+              # it timestamps, so we don't.
+              let response = defaultRegistry.toText(showTimestamp = false)
+              let headers = HttpTable.init([("Content-Type", CONTENT_TYPE)])
+              return await request.respond(Http200, response, headers)
+          elif request.uri.path == "/health":
+            return await request.respond(Http200, "OK")
+          else:
+            return await request.respond(Http404, "Try /metrics")
+        except CatchableError as exc:
+          printError(exc.msg)
+
+    let
+      http =
+        block:
+          let socketFlags = {ServerFlags.ReuseAddr}
+          let res = HttpServerRef.new(server.address, cb,
+                                      socketFlags = socketFlags)
+          if res.isErr():
+            lastError = res.error()
+            nil
+          else:
+            res.get()
+      reqTransp =
+        try:
+          fromPipe(server.requestPipe.read)
+        except CatchableError as exc:
+          await http.closeWait()
+          return
+      respTransp =
+        try:
+          fromPipe(server.responsePipe.write)
+        except CatchableError as exc:
+          await http.closeWait()
+          await reqTransp.closeWait()
+          return
+      threadData = MetricsThreadData(
+        reqTransp: reqTransp, respTransp: respTransp, http: http)
+
+    while true:
+      let res = await asyncStep(server, threadData, lastError)
+      if not(res):
+        break
+
+    await allFutures(reqTransp.closeWait(), respTransp.closeWait())
+
+  proc serveMetricsServer(server: MetricsServerData) {.thread.} =
+    ignoreSignalsInThread()
+    let loop {.used.} = getThreadDispatcher()
+    try:
+      waitFor asyncLoop(server)
+    except Exception as exc:
+      printError($exc.msg)
 
 proc startMetricsHttpServer*(address = "127.0.0.1", port = Port(8000)) {.
-     raises: [Exception].} =
+     raises: [Exception],
+     deprecated: "Please use MetricsHttpServerRef API".} =
   when defined(metrics):
     httpServerThread.createThread(serveHttp, initTAddress(address, port))
 
+proc new*(t: typedesc[MetricsHttpServerRef], address: string,
```

Comment:
this looks more like a "general" feature to run a chronos event loop on a different thread - it's not really tied to metrics - does it make sense to move such a feature to chronos instead?

Reply:
More "generic" solution should be found to put it in `chronos`. For example here i have used response bound to 256 bytes, but for `chronos` its not suitable, some procedures could return bigger responses. But yeah at some point similar solution will be added to `chronos`.